File: 01_C++17 和 python 中好用的文件操作/main.py
import os
import shutil

def eg1_1():
    """判断是否为绝对路径"""
    # abs_path = "C:\\Users"
    abs_path = "/usr/local"
    print("os.path.isabs({}) : {}".format(abs_path, os.path.isabs(abs_path)))
    rel_path = "../"
    print("os.path.isabs({}) : {}".format(rel_path, os.path.isabs(rel_path)))

def eg1_2():
    """路径分割"""
    path = "../test_dir/1.txt"
    print("os.path.split(path) : {}".format(os.path.split(path)))
    print("os.path.dirname(path) : {}".format(os.path.dirname(path)))
    print("os.path.basename(path) : {}".format(os.path.basename(path)))

def eg1_3():
    """路径拼接"""
    path = "../test_dir"
    txt1_path = os.path.join(path, "1.txt")
    txt2_path = os.path.join(path, "1_dir", "2.txt")
    print("txt1_path : {}".format(txt1_path))
    print("txt2_path : {}".format(txt2_path))

def eg2_1():
    """获取当前路径"""
    current_path = os.getcwd()
    print("current_path : {}".format(current_path))

def eg2_2():
    """返回指定目录下的所有文件/文件夹"""
    for i in os.listdir("./test_dir"):
        print(i)

def eg2_3():
    """递归返回指定目录下的所有文件/文件夹"""
    for root, dirs, files in os.walk("./test_dir"):
        print("root : {}, dirs : {}, files : {}".format(root, dirs, files))

def eg2_4():
    """判断路径是否存在"""
    path = "C:\\Users"
    print("os.path.exists({}) : {}".format(path, os.path.exists(path)))

def eg2_5():
    """判断路径是文件还是目录"""
    file_path = "./test_dir/1.txt"
    dir_path = "./test_dir/1_dir"
    print("os.path.isfile({}) : {}".format(file_path, os.path.isfile(file_path)))
    print("os.path.isdir({}) : {}".format(dir_path, os.path.isdir(dir_path)))

def eg2_6():
    """返回绝对路径"""
    path = "../../-PyTorch-"
    abs_path = os.path.abspath(path)
    print("abs_path : {}".format(abs_path))
    print("os.path.exists({}) : {}".format(abs_path, os.path.exists(abs_path)))

def eg3_1():
    """文件拷贝"""
    print("~~~~~~before copy_file~~~~~~")
    for i in os.listdir("./test_dir"):
        print(i)
    src_path = "./test_dir/1.txt"
    dst_path = "./test_dir/eg3_1.txt"
    shutil.copyfile(src_path, dst_path)
    print("~~~~~~after copy_file~~~~~~")
    for i in os.listdir("./test_dir"):
        print(i)

def eg3_2():
    """文件删除"""
    print("~~~~~~before remove~~~~~~")
    for i in os.listdir("./test_dir"):
        print(i)
    rm_path = "./test_dir/eg3_1.txt"
    os.remove(rm_path)
    print("~~~~~~after remove~~~~~~")
    for i in os.listdir("./test_dir"):
        print(i)

def eg3_3():
    """路径拷贝"""
    print("~~~~~~before copy[1_dir]~~~~~~")
    for i in os.listdir("./test_dir/1_dir"):
        print(i)
    src_path = "./test_dir/1_dir"
    dst_path = "./test_dir/eg3_3_dir"
    shutil.copytree(src_path, dst_path)
    print("~~~~~~after copy[eg3_3_dir]~~~~~~")
    for i in os.listdir("./test_dir/eg3_3_dir"):
        print(i)

def eg3_4():
    """路径删除"""
    dir_path = "./test_dir/eg3_3_dir"
    print("os.path.exists({}) : {}".format(dir_path, os.path.exists(dir_path)))
    shutil.rmtree(dir_path)
    print("os.path.exists({}) : {}".format(dir_path, os.path.exists(dir_path)))

if __name__ == "__main__":
    # `filesystem::path` vs. `os.path`
    # eg1_1()
    # eg1_2()
    # eg1_3()

    # `filesystem` vs. `os`
    # eg2_1()
    # eg2_2()
    # eg2_3()
    # eg2_4()
    # eg2_5()
    # eg2_6()

    # `filesystem` vs. `shutil`
    # eg3_1()
    # eg3_2()
    eg3_3()
    # eg3_4()
    #
    # print("~~~~~~下课~~~~~~")

File: 01_C++17 和 python 中好用的文件操作/main.cpp
#include <filesystem>
#include <iostream>

void eg1_1() {
    /*判断是否为绝对路径*/
    // std::filesystem::path abs_path = "C:\\Users";
    std::filesystem::path abs_path = "/usr/local";  // 注意，实例化path的时候可以直接用等号
    std::cout << "abs_path.is_absolute() : " << abs_path.is_absolute() << std::endl;
    std::cout << "abs_path.is_relative() : " << abs_path.is_relative() << std::endl;

    std::filesystem::path rel_path = "../";
    std::cout << "rel_path.is_absolute() : " << rel_path.is_absolute() << std::endl;
    std::cout << "rel_path.is_relative() : " << rel_path.is_relative() << std::endl;
}

void eg1_2() {
    /*路径分割*/
    std::filesystem::path path = "../test_dir/1.txt";
    std::cout << "path.relative_path() : " << path.relative_path() << std::endl;
    std::cout << "path.parent_path() : " << path.parent_path() << std::endl;
    std::cout << "path.filename() : " << path.filename() << std::endl;
    std::cout << "path.stem() : " << path.stem() << std::endl;
    std::cout << "path.extension() : " << path.extension() << std::endl;
}

void eg1_3() {
    /*路径拼接*/
    std::filesystem::path path = "../test_dir";
    std::filesystem::path txt1_path = path / "1.txt";
    std::filesystem::path txt2_path = path / "1_dir" / "2.txt";
    std::cout << "txt1_path : " << txt1_path << std::endl;
    std::cout << "txt2_path : " << txt2_path << std::endl;
    path /= "1.txt";
    std::cout << "path : " << path << std::endl;
}

void eg2_1() {
    /*获取当前路径*/
    std::filesystem::path current_path = std::filesystem::current_path();
    std::cout << "current_path : " << current_path << std::endl;
}

void eg2_2() {
    /*返回指定目录下的所有文件/文件夹*/
    std::filesystem::directory_iterator iter("../test_dir");
    for(const auto &i : iter) {
        std::cout << i.path() <<std::endl;
    }
}

void eg2_3() {
    /*递归返回指定目录下的所有文件/文件夹*/
    std::filesystem::recursive_directory_iterator iter("../test_dir");
    for(const auto &i : iter) {
        std::cout << i.path() <<std::endl;
    }
}

void eg2_4() {
    /*判断路径是否存在*/
    bool exist = std::filesystem::exists("C:\\Users");
    std::cout << "exist : " << exist << std::endl;
}

void eg2_5() {
    /*判断路径是文件还是目录*/
    std::filesystem::path file_path = "../test_dir/1.txt";
    std::filesystem::path dir_path = "../test_dir/1_dir";
    std::cout << "is_regular_file(file_path) : "
              << std::filesystem::is_regular_file(file_path)
              << std::endl;
    std::cout << "is_directory(dir_path) : "
              << std::filesystem::is_directory(dir_path)
              << std::endl;
}

void eg2_6() {
    /*返回绝对路径*/
    std::filesystem::path path = "../test_dir";
    std::filesystem::path abs_path = std::filesystem::absolute(path);
    std::cout << "abs_path : " << abs_path << std::endl;
    std::cout << "exists(abs_path) : " << std::filesystem::exists(abs_path) << std::endl;
}

void eg3_1() {
    /*文件拷贝*/
    std::cout << "~~~~~~before copy_file~~~~~~" << std::endl;
    for (const auto &i :std::filesystem::directory_iterator("../test_dir")) {
        std::cout << i.path() << std::endl;
    }
    std::filesystem::copy_file("../test_dir/1.txt", "../test_dir/eg3_1.txt");
    std::cout << "~~~~~~after copy_file~~~~~~" << std::endl;
    for (const auto &i :std::filesystem::directory_iterator("../test_dir")) {
        std::cout << i.path() << std::endl;
    }
}

void eg3_2() {
    /*文件删除*/
    std::cout << "~~~~~~before remove~~~~~~" << std::endl;
    for (const auto &i :std::filesystem::directory_iterator("../test_dir")) {
        std::cout << i.path() << std::endl;
    }
    std::filesystem::remove("../test_dir/eg3_1.txt");
    std::cout << "~~~~~~after remove~~~~~~" << std::endl;
    for (const auto &i :std::filesystem::directory_iterator("../test_dir")) {
        std::cout << i.path() << std::endl;
    }
}

void eg3_3() {
    /*路径拷贝*/
    std::cout << "~~~~~~before copy[1_dir]~~~~~~" << std::endl;
    for (const auto &i :std::filesystem::directory_iterator("../test_dir/1_dir")) {
        std::cout << i.path() << std::endl;
    }
    std::filesystem::copy("../test_dir/1_dir",
                          "../test_dir/eg3_3_dir",
                          std::filesystem::copy_options::recursive);
    std::cout << "~~~~~~after copy[eg3_3_dir]~~~~~~" << std::endl;
    for (const auto &i :std::filesystem::directory_iterator("../test_dir/eg3_3_dir")) {
        std::cout << i.path() << std::endl;
    }
}

void eg3_4() {
    /*递归删除*/
    std::filesystem::path dir_path = "../test_dir/eg3_3_dir";
    std::cout << "exists(dir_path) : " << std::filesystem::exists(dir_path) << std::endl;
    std::filesystem::remove_all(dir_path);
    std::cout << "exists(dir_path) : " << std::filesystem::exists(dir_path) << std::endl;
}

int main(int argc, char* argv[]) {
    // `filesystem::path` vs. `os.path`
//    eg1_1();
//    eg1_2();
//    eg1_3();

    // `filesystem` vs. `os`
//    eg2_1();
//    eg2_2();
//    eg2_3();
//    eg2_4();
//    eg2_5();
//    eg2_6();

    // `filesystem` vs. `shutil`
//    eg3_1();
//    eg3_2();
//    eg3_3();
//    eg3_4();

    return 0;
}